File: csharp/Tdcv2/Cli/Pack.cs
using System.Globalization;
using System.Text;
using Tdcv2.Packs;

namespace Tdcv2.Cli;

/// <summary>
/// <c>tdcv2 pack</c> — the data packs, from the shared registry.
/// </summary>
/// <remarks>
/// <para>
/// The library ships with English and the USA and nothing else, because the packs are meant to grow
/// to a size no library should carry. Everything else is downloaded on demand from one registry that
/// every implementation reads, so a pack fetched by any of them works in the others.
/// </para>
/// <para>
/// A bundle is axis-pure — one language (<c>en</c>), or one country (<c>usa</c>), or <c>common</c> —
/// because language and country are independent and compose: US English is common + en + usa.
/// Everything lands in ONE tree, at its address path: <c>&lt;store&gt;/en/…</c>,
/// <c>&lt;store&gt;/countries/usa/…</c>. Which bundle owns which path is written down in
/// <c>&lt;store&gt;/.tdcv2-installed.json</c> rather than implied by a folder name, so ten languages
/// and a hundred countries are one folder and one <c>dataPaths</c> entry.
/// </para>
/// <para>
/// Where things go is decided by the config cascade, not guessed: <c>packStore</c> from the nearest
/// <c>tdcv2.config.json</c>, else the global one. With neither, <c>init</c> has not been run and this
/// says so rather than inventing a folder.
/// </para>
/// </remarks>
public static class Pack
{
    private const string Usage = @"Usage: tdcv2 pack [command]

  list                  Show what can be installed, and what already is
  add <id>...           Download and install one or more bundles
  remove <id>...        Uninstall, and drop them from the config

  --registry <url>      Use another registry (default: the public one)
";

    /// <summary>Where downloads go, and which config file records them.</summary>
    public sealed record Store(string Path, string ConfigPath);

    /// <summary>
    /// The pack folder and the config that owns it, from the cascade.
    /// </summary>
    /// <remarks>
    /// A project config wins over the global one — packs belong to the project that uses them.
    /// Neither means <c>init</c> has not run, and guessing a folder here would put data somewhere the
    /// next command would not look.
    /// </remarks>
    public static Store ResolveStore(string cwd)
    {
        ProjectConfig.Resolved resolved = ProjectConfig.Load(cwd);

        string? configPath = ProjectConfig.FindProjectConfig(cwd)
            ?? ProjectConfig.GlobalConfigPath();
        if (configPath is null || !File.Exists(configPath))
        {
            throw new PackRegistry.PackException(
                "no tdcv2.config.json found — run `tdcv2 init` first");
        }

        return resolved.PackStore is null
            ? throw new PackRegistry.PackException(
                $"config \"{configPath}\" does not name a \"packStore\" — add one, or re-run "
                + "`tdcv2 init --force`")
            : new Store(resolved.PackStore, configPath);
    }

    public static int Run(
        IReadOnlyList<string> argv, string cwd, TextWriter stdout, TextWriter stderr)
    {
        if (argv.Contains("-h") || argv.Contains("--help"))
        {
            stdout.Write(Usage);
            return 0;
        }

        string? registryUrl = null;
        var rest = new List<string>();
        for (int i = 0; i < argv.Count; i++)
        {
            string arg = argv[i];
            if (arg == "--registry")
            {
                i++;
                if (i >= argv.Count)
                {
                    stderr.Write("tdcv2: missing value for --registry\n");
                    return 2;
                }

                registryUrl = argv[i];
            }
            else if (arg.StartsWith("--registry=", StringComparison.Ordinal))
            {
                registryUrl = arg["--registry=".Length..];
            }
            else
            {
                rest.Add(arg);
            }
        }

        // No subcommand on a terminal opens the picker; anywhere else — a pipe, a script — it
        // lists, which answers the same question without needing a keyboard.
        bool bare = rest.Count == 0;
        string command = rest.Count == 0 ? "list" : rest[0];
        List<string> ids = rest.Count == 0 ? new List<string>() : rest.Skip(1).ToList();

        Store store;
        try
        {
            store = ResolveStore(cwd);
        }
        catch (PackRegistry.PackException e)
        {
            stderr.Write("tdcv2: " + e.Message + "\n");
            return 2;
        }

        var registry = registryUrl is null ? new PackRegistry() : new PackRegistry(registryUrl);

        try
        {
            // Before anything reads the store: a store from an older tdcv2 is in the old
            // per-bundle layout, which `list`, `add` and `remove` all now misread. The first
            // `tdcv2 pack` after an upgrade fixes it, once, and says what it did.
            PackStore.StoreMigration? migration =
                PackStore.Migrate(store.Path, store.ConfigPath);
            if (migration is not null)
            {
                ReportMigration(migration, stderr);
            }

            if (bare && PackPicker.Usable())
            {
                return Browse(registry, store, stdout, stderr);
            }

            switch (command)
            {
                case "list":
                    return List(registry, store, stdout);
                case "add":
                    if (ids.Count == 0)
                    {
                        stderr.Write("tdcv2: `pack add` needs at least one bundle id\n");
                        return 2;
                    }

                    return Add(registry, store, ids, stdout, stderr);
                case "remove":
                    if (ids.Count == 0)
                    {
                        stderr.Write("tdcv2: `pack remove` needs at least one bundle id\n");
                        return 2;
                    }

                    return Remove(store, ids, stdout, stderr);
                default:
                    stderr.Write(
                        $"tdcv2: unknown pack command \"{command}\" (use list | add | remove)\n");
                    return 2;
            }
        }
        catch (Exception e) when (e is PackRegistry.PackException or IOException
                                      or ProjectConfig.ConfigException)
        {
            stderr.Write("tdcv2: " + e.Message + "\n");
            return 2;
        }
    }

    /// <summary>Where a description starts: two spaces, the id column, a space.</summary>
    /// <summary>The narrowest the id column ever gets: two spaces, twelve, a space.</summary>
    /// <remarks>
    /// It grows to fit the widest id in the catalogue — <c>sao_tome_and_principe</c> is twenty-one
    /// characters, and a fixed twelve pushed every column after it out of line on that row alone.
    /// It never shrinks below this, so a short catalogue keeps the shape the shared CLI fixture
    /// pins.
    /// </remarks>
    private const int MinIdWidth = 12;

    /// <summary>
    /// Text folded to the terminal, with every line under the same indent.
    /// </summary>
    /// <remarks>
    /// Descriptions run to two hundred characters. Printed as one line each they wrap wherever the
    /// window happens to end and the remainder lands hard against the left margin, which turns a list
    /// of a hundred packs into a wall nobody can read down. Off a terminal there is no width to ask
    /// for, so 80 is assumed — the same answer every implementation gives, which keeps their output
    /// identical when it is piped.
    /// </remarks>
    private static IReadOnlyList<string> Wrap(string text, int indent)
    {
        int width = Math.Max(40, TerminalColumns()) - indent;
        string pad = new(' ', indent);
        var lines = new List<string>();
        var line = new StringBuilder();
        foreach (string word in text.Trim().Split(
                     (char[]?)null, StringSplitOptions.RemoveEmptyEntries))
        {
            if (line.Length == 0)
            {
                line.Append(word);
            }
            else if (line.Length + 1 + word.Length <= width)
            {
                line.Append(' ').Append(word);
            }
            else
            {
                lines.Add(pad + line);
                line = new StringBuilder(word);
            }
        }

        if (line.Length > 0)
        {
            lines.Add(pad + line);
        }

        return lines;
    }

    /// <summary>
    /// How wide the terminal is, or 80 when there is not one.
    /// </summary>
    /// <remarks>
    /// Asked of the terminal itself, never of <c>COLUMNS</c>: the other implementations ask the file
    /// descriptor, and an environment variable that only one of them honours is a way for the same
    /// command to print differently in the same shell.
    /// </remarks>
    private static int TerminalColumns()
    {
        try
        {
            return Console.IsOutputRedirected ? 80 : Console.WindowWidth;
        }
        catch (IOException)
        {
            return 80;
        }
    }

    /// <summary>The picker, and then whatever it decided.</summary>
    /// <remarks>
    /// The picker draws and returns; installing and removing stay here, so the download progress,
    /// the digests and the config writing live in one place rather than two.
    /// </remarks>
    private static int Browse(
        PackRegistry registry, Store store, TextWriter stdout, TextWriter stderr)
    {
        PackRegistry.Index index = registry.ReadIndex();
        if (index.Bundles.Count == 0)
        {
            stdout.Write("No bundles in the registry.\n");
            return 0;
        }

        var here = PackRegistry.Installed(store.Path).ToHashSet(StringComparer.Ordinal);
        PackPicker.Decision? decision = PackPicker.Run(index.Bundles, here, stdout);
        if (decision is null || (decision.Install.Count == 0 && decision.Remove.Count == 0))
        {
            stdout.Write("Nothing changed.\n");
            return 0;
        }

        if (decision.Remove.Count > 0)
        {
            Remove(store, decision.Remove, stdout, stderr);
        }

        return decision.Install.Count > 0
            ? Add(registry, store, decision.Install, stdout, stderr)
            : 0;
    }

    private static int List(PackRegistry registry, Store store, TextWriter stdout)
    {
        PackRegistry.Index index = registry.ReadIndex();
        var here = PackRegistry.Installed(store.Path).ToHashSet(StringComparer.Ordinal);

        if (index.Bundles.Count == 0)
        {
            stdout.Write("No bundles in the registry.\n");
            return 0;
        }

        int idWidth = Math.Max(MinIdWidth, index.Bundles.Max(b => b.Id.Length));
        // Two spaces of margin, the id column, one space — where a description and the
        // "installed" mark both start.
        int indent = 2 + idWidth + 1;

        stdout.Write("Available data packs:\n");
        stdout.Write("\n");
        foreach (PackRegistry.Bundle bundle in index.Bundles)
        {
            string mark = here.Contains(bundle.Id) ? "✓ installed" : " ";
            stdout.Write(string.Create(
                CultureInfo.InvariantCulture,
                $"  {bundle.Id.PadRight(idWidth)} {mark,-12} {bundle.Name} ({Megabytes(bundle.Bytes)})\n"));
            foreach (string line in Wrap(bundle.Description, indent))
            {
                stdout.Write(line + "\n");
            }

            stdout.Write("\n");
        }

        // An id in the store that the registry no longer lists still works; saying so beats a silent
        // omission that reads like the pack is gone.
        var known = index.Bundles.Select(b => b.Id).ToHashSet(StringComparer.Ordinal);
        List<string> orphans = here
            .Where(id => !known.Contains(id))
            .OrderBy(id => id, StringComparer.Ordinal)
            .ToList();
        if (orphans.Count > 0)
        {
            stdout.Write("Installed but not in this registry: " + string.Join(", ", orphans) + "\n");
            stdout.Write("\n");
        }

        stdout.Write("Install with: tdcv2 pack add <id>\n");
        return 0;
    }

    private static int Add(
        PackRegistry registry, Store store, IReadOnlyList<string> ids, TextWriter stdout,
        TextWriter stderr)
    {
        PackRegistry.Index index = registry.ReadIndex();
        // Every id is looked up before anything is downloaded, so a typo in the third one does not
        // leave the first two half-installed.
        List<PackRegistry.Bundle> bundles = ids.Select(index.Find).ToList();

        foreach (PackRegistry.Bundle bundle in bundles)
        {
            stderr.Write($"tdcv2: downloading {bundle.Id} ({Megabytes(bundle.Bytes)})…\n");
            PackRegistry.Installation result = registry.Install(bundle, store.Path);

            // The STORE goes into the config, once, however many bundles land in it — a bundle
            // no longer has a folder of its own to register.
            string stored = ProjectConfig.StoredPath(store.ConfigPath, store.Path);
            bool added = ProjectConfig.Register(store.ConfigPath, new[] { store.Path });
            stdout.Write(
                $"Installed {bundle.Id}: {result.Files} files → "
                + $"{InstalledAt(store.Path, result.Paths)}\n");
            stdout.Write(
                added
                    ? $"  registered {stored} in {store.ConfigPath}\n"
                    : $"  already registered in {store.ConfigPath}\n");
        }

        return 0;
    }

    /// <summary>Where a bundle's data ended up, for the line that reports an install.</summary>
    private static string InstalledAt(string store, IReadOnlyList<string> paths) =>
        string.Join(", ", paths.Select(p => PackStore.Resolve(store, p)));

    /// <summary>
    /// Uninstall bundles: delete exactly the paths the store recorded for each, and drop the store
    /// from the config once nothing is left in it. No network.
    /// </summary>
    /// <remarks>
    /// Deleting by record rather than by folder name is what a shared tree costs and what it buys:
    /// <c>russia</c> lives at <c>countries/russia</c> beside <c>countries/usa</c>, and only the
    /// recorded path goes. Because installs SHADOW the bundled default rather than replacing it,
    /// removing a bundle simply lets the default resurface — no hole.
    /// </remarks>
    private static int Remove(
        Store store, IReadOnlyList<string> ids, TextWriter stdout, TextWriter stderr)
    {
        PackStore.InstalledRecord record = PackStore.Read(store.Path);
        foreach (string id in ids)
        {
            PackStore.InstalledBundle? entry = record.Bundles.FirstOrDefault(b => b.Id == id);
            if (entry is null)
            {
                // Not an error. `remove` is asked for to reach a state, and that state already
                // holds; exiting non-zero would make an idempotent script fail on its second run.
                stderr.Write($"tdcv2: \"{id}\" is not installed — nothing to remove\n");
                continue;
            }

            PackStore.DeleteOwnedPaths(store.Path, entry.Paths);
            record = PackStore.Without(record, id);
            PackStore.Write(store.Path, record);
            stdout.Write($"Removed {id} ({InstalledAt(store.Path, entry.Paths)})\n");
        }

        // The store stays registered while it still holds something: one entry serves every
        // bundle, so it only comes out when the last one does.
        if (record.Bundles.Count == 0
            && ProjectConfig.Unregister(store.ConfigPath, new[] { store.Path }))
        {
            stdout.Write(
                $"  store now empty — unregistered {store.Path} from {store.ConfigPath}\n");
        }

        return 0;
    }

    /// <summary>
    /// Say what the move to the flat layout did.
    /// </summary>
    /// <remarks>
    /// On stderr: <c>pack list</c> prints a catalogue people pipe, and a one-off notice about the
    /// store is not part of it.
    /// </remarks>
    private static void ReportMigration(PackStore.StoreMigration migration, TextWriter stderr)
    {
        var lines = new List<string>
        {
            $"tdcv2: pack store \"{migration.Store}\" used the old per-bundle layout; "
            + "moved it to the flat one.",
        };
        foreach (PackStore.StoreMove move in migration.Moves)
        {
            lines.Add(
                $"  {move.Id}: {move.From} → {string.Join(", ", move.To)} ({move.Files} files)");
        }

        if (migration.DroppedDataPaths > 0)
        {
            lines.Add($"  dropped {migration.DroppedDataPaths} per-bundle dataPaths entries");
        }

        if (migration.Registered is not null)
        {
            lines.Add($"  registered {migration.Registered} instead");
        }

        if (migration.Leftovers.Count > 0)
        {
            string shown = string.Join(", ", migration.Leftovers.Take(3));
            string more = migration.Leftovers.Count > 3
                ? $", … and {migration.Leftovers.Count - 3} more"
                : "";
            lines.Add($"  left where they were (not pack data): {shown}{more}");
        }

        stderr.Write(string.Join("\n", lines) + "\n");
    }

    private static string Megabytes(long bytes) =>
        (bytes / 1024.0 / 1024.0).ToString("F1", CultureInfo.InvariantCulture) + " MB";
}

File: python/src/tdcv2/cli/pack.py
"""``tdcv2 pack`` — the data packs, from the shared registry.

The library ships with English and the USA and nothing else, because the packs are meant to grow to
a size no library should carry. Everything else is downloaded on demand from one registry that all
three implementations read, so a pack fetched by any of them works in the other two.

A bundle is axis-pure — one language (``en``), or one country (``usa``), or ``common`` — because
language and country are independent and compose: US English is common + en + usa. Everything lands
in ONE tree, at its address path: ``<store>/ru/…``, ``<store>/countries/russia/…``. Which bundle
owns which path is written down in ``<store>/.tdcv2-installed.json`` rather than implied by a folder
name, so ten languages and a hundred countries are one folder and one ``dataPaths`` entry.

Where things go is decided by the config cascade, not guessed: ``packStore`` from the nearest
``tdcv2.config.json``, else the global one. With neither, `init` has not been run and this says so
rather than inventing a folder.
"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from pathlib import Path

from ..packs import project_config
from ..packs.registry import DEFAULT_BASE_URL, Registry
from ..packs.store import (
    PackError,
    StoreMigration,
    delete_owned_paths,
    find_bundle,
    installed,
    migrate_store,
    read_installed,
    without_bundle,
    write_installed,
)

USAGE = """Usage: tdcv2 pack [command]

  list                  Show what can be installed, and what already is
  add <id>...           Download and install one or more bundles
  remove <id>...        Uninstall, and drop them from the config

  --registry <url>      Use another registry (default: the public one)
"""


@dataclass(frozen=True, slots=True)
class Store:
    """Where downloads go, and which config file records them."""

    path: Path
    config_path: Path


def resolve_store(cwd: Path | None = None) -> Store:
    """The pack folder and the config that owns it, from the cascade.

    A project config wins over the global one — packs belong to the project that uses them. Neither
    means `init` has not run, and guessing a folder here would put data somewhere the next command
    would not look.
    """
    here = (cwd or Path.cwd()).resolve()
    resolved = project_config.load(here)

    config_path = project_config.find_project_config(here)
    if config_path is None:
        config_path = project_config.global_config_path()
    if config_path is None or not config_path.is_file():
        raise PackError("no tdcv2.config.json found — run `tdcv2 init` first")

    if resolved.pack_store is None:
        raise PackError(
            f'config "{config_path}" does not name a "packStore" — '
            "add one, or re-run `tdcv2 init --force`"
        )
    return Store(resolved.pack_store, config_path)


def run_pack(argv: list[str], cwd: Path | None = None) -> int:
    if any(a in ("-h", "--help") for a in argv):
        sys.stdout.write(USAGE)
        return 0

    registry_url = DEFAULT_BASE_URL
    rest: list[str] = []
    i = 0
    while i < len(argv):
        arg = argv[i]
        if arg == "--registry":
            i += 1
            if i >= len(argv):
                sys.stderr.write("tdcv2: missing value for --registry\n")
                return 2
            registry_url = argv[i]
        elif arg.startswith("--registry="):
            registry_url = arg[len("--registry=") :]
        else:
            rest.append(arg)
        i += 1

    # No subcommand and a real terminal on both ends: browse the catalogue instead of printing it.
    # Anywhere else — a pipe, a script, CI — printing is the only sensible thing to do.
    # TDCV2_NO_PICKER turns the picker off deliberately: for a script that wants the printed list,
    # and for anyone who would rather not be dropped into a full-screen program.
    interactive = (
        not rest
        and not os.environ.get("TDCV2_NO_PICKER")
        and sys.stdin.isatty()
        and sys.stdout.isatty()
    )
    command = rest[0] if rest else "list"
    ids = rest[1:]

    try:
        store = resolve_store(cwd)
    except PackError as e:
        sys.stderr.write(f"tdcv2: {e}\n")
        return 2

    try:
        # Before anything reads the store: a store from an older tdcv2 is in the old per-bundle
        # layout, which `list`, `add` and `remove` all now misread. The first `tdcv2 pack` after
        # an upgrade fixes it, once, and says what it did.
        migration = migrate_store(store.path, store.config_path)
        if migration is not None:
            _report_migration(migration)

        if interactive:
            return _browse(Registry(registry_url), store)
        if command == "list":
            return _list(Registry(registry_url), store)
        if command == "add":
            if not ids:
                sys.stderr.write("tdcv2: `pack add` needs at least one bundle id\n")
                return 2
            return _add(Registry(registry_url), store, ids)
        if command == "remove":
            if not ids:
                sys.stderr.write("tdcv2: `pack remove` needs at least one bundle id\n")
                return 2
            return _remove(store, ids)
    except PackError as e:
        sys.stderr.write(f"tdcv2: {e}\n")
        return 2

    sys.stderr.write(f'tdcv2: unknown pack command "{command}" (use list | add | remove)\n')
    return 2


#: Where a description starts: two spaces, the id column, a space.
# The narrowest the id column ever gets: two spaces, twelve, a space. It grows to fit the widest
# id in the catalogue — "sao_tome_and_principe" is twenty-one characters, and a fixed twelve
# pushed every column after it out of line on that row alone. It never shrinks below this, so a
# short catalogue keeps the shape the shared CLI fixture pins.
_MIN_ID_WIDTH = 12


def _wrap(text: str, indent: int) -> list[str]:
    """Text folded to the terminal, with every line under the same indent.

    Descriptions run to two hundred characters. Printed as one line each they wrap wherever the
    window happens to end and the remainder lands hard against the left margin, which turns a list
    of a hundred packs into a wall nobody can read down. Off a terminal there is no width to ask
    for, so 80 is assumed — the same answer every implementation gives, which keeps their output
    identical when it is piped.
    """
    try:
        columns = os.get_terminal_size().columns
    except OSError:
        columns = 80
    width = max(40, columns) - indent
    pad = " " * indent
    lines: list[str] = []
    line = ""
    for word in text.split():
        if not line:
            line = word
        elif len(line) + 1 + len(word) <= width:
            line += " " + word
        else:
            lines.append(pad + line)
            line = word
    if line:
        lines.append(pad + line)
    return lines


def _browse(registry: Registry, store: Store) -> int:
    """The catalogue, browsed rather than listed.

    The picker only decides; installing and removing stay here, so the digests, the progress
    reporting and the config writing have one home whether the ids were typed or pointed at.
    """
    from .pack_picker import run_picker

    index = registry.index()
    if not index.bundles:
        sys.stdout.write("No bundles in the registry.\n")
        return 0

    decision = run_picker(index.bundles, set(installed(store.path)))
    if decision is None:
        return 0

    if decision.remove:
        _remove(store, decision.remove)
    if decision.install:
        return _add(registry, store, decision.install)
    if not decision.remove:
        sys.stdout.write("Nothing selected.\n")
    return 0


def _list(registry: Registry, store: Store) -> int:
    index = registry.index()
    here = set(installed(store.path))

    if not index.bundles:
        sys.stdout.write("No bundles in the registry.\n")
        return 0

    id_width = max(_MIN_ID_WIDTH, *(len(b.id) for b in index.bundles))
    # Two spaces of margin, the id column, one space — where a description and the "installed"
    # mark both start.
    indent = 2 + id_width + 1

    sys.stdout.write("Available data packs:\n\n")
    for bundle in index.bundles:
        mark = "\u2713 installed" if bundle.id in here else " "
        sys.stdout.write(
            f"  {bundle.id:<{id_width}} {mark:<12} "
            f"{bundle.name} ({bundle.bytes / 1048576:.1f} MB)\n"
        )
        for line in _wrap(bundle.description, indent):
            sys.stdout.write(line + "\n")
        sys.stdout.write("\n")

    # An id in the store that the registry no longer lists still works; saying so beats a silent
    # omission that reads like the pack is gone.
    orphans = sorted(here - {b.id for b in index.bundles})
    if orphans:
        sys.stdout.write("Installed but not in this registry: " + ", ".join(orphans) + "\n\n")

    sys.stdout.write("Install with: tdcv2 pack add <id>\n")
    return 0


def _installed_at(store: Path, paths: list[str]) -> str:
    """Where a bundle's data ended up, for the line that reports an install or a removal."""
    return ", ".join(str(store / path) for path in paths)


def _add(registry: Registry, store: Store, ids: list[str]) -> int:
    index = registry.index()
    # Every id is looked up before anything is downloaded, so a typo in the third one does not
    # leave the first two half-installed.
    bundles = [index.find(bundle_id) for bundle_id in ids]
    for bundle in bundles:
        sys.stderr.write(f"tdcv2: downloading {bundle.id} ({bundle.bytes / 1048576:.1f} MB)…\n")
        result = registry.install(bundle, store.path)
        # The STORE goes into the config, once, however many bundles land in it \u2014 not each one.
        stored = project_config.storable(store.config_path, store.path)
        added = project_config.register(store.config_path, [store.path])
        sys.stdout.write(
            f"Installed {bundle.id}: {result.files} files "
            f"\u2192 {_installed_at(store.path, result.paths)}\n"
            + (
                f"  registered {stored} in {store.config_path}\n"
                if added
                else f"  already registered in {store.config_path}\n"
            )
        )
    return 0


def _remove(store: Store, ids: list[str]) -> int:
    """Bundles uninstalled: exactly the paths the store recorded for each, and nothing beside them.

    Deleting by record rather than by folder name is what a shared tree costs and what it buys:
    ``russia`` lives at ``countries/russia`` next to ``countries/usa``, and only the recorded path
    goes. Because installs SHADOW the bundled default rather than replacing it, removing a bundle
    simply lets the default resurface — no hole. Nothing here touches the network.
    """
    record = read_installed(store.path)
    for bundle_id in ids:
        entry = find_bundle(record, bundle_id)
        if entry is None:
            # Not an error. `remove` is asked for to reach a state, and that state already holds;
            # exiting non-zero would make an idempotent script fail on its second run.
            sys.stderr.write(f'tdcv2: "{bundle_id}" is not installed — nothing to remove\n')
            continue
        delete_owned_paths(store.path, entry.paths)
        record = without_bundle(record, bundle_id)
        write_installed(store.path, record)
        sys.stdout.write(f"Removed {bundle_id} ({_installed_at(store.path, entry.paths)})\n")

    # The store stays registered while it still holds something: one entry serves every bundle in
    # it, so it only comes out when the last one leaves.
    if not record.bundles and project_config.unregister(store.config_path, [store.path]):
        sys.stdout.write(
            f"  store now empty — unregistered {store.path} from {store.config_path}\n"
        )
    return 0


def _report_migration(migration: StoreMigration) -> None:
    """A store moved to the flat layout, said out loud.

    On stderr: `pack list` prints a catalogue people pipe, and a one-off notice about the store is
    no part of it.
    """
    lines = [
        f'tdcv2: pack store "{migration.store}" used the old per-bundle layout; '
        "moved it to the flat one."
    ]
    for move in migration.moves:
        lines.append(f"  {move.id}: {move.source} → {', '.join(move.to)} ({move.files} files)")
    if migration.dropped_data_paths > 0:
        lines.append(f"  dropped {migration.dropped_data_paths} per-bundle dataPaths entries")
    if migration.registered is not None:
        lines.append(f"  registered {migration.registered} instead")
    if migration.leftovers:
        shown = ", ".join(migration.leftovers[:3])
        rest = len(migration.leftovers) - 3
        more = f", … and {rest} more" if rest > 0 else ""
        lines.append(f"  left where they were (not pack data): {shown}{more}")
    sys.stderr.write("\n".join(lines) + "\n")
